### PR TITLE
Sync up Agent and API's renewers.

### PR DESCRIPTION
**Pull request by jefferai**
This introduces a new type, LifetimeWatcher, which can handle both
renewable and non-renewable secrets, modeled after the version in Agent.
It allows the user to select behavior, with the new style being the
default when calling Start(), and old style if using the legacy Renew()
call.

This also unifies the logic in the lease/auth renew calls, which were (other than some variables) otherwise exact copypasta.

No tests have been modified (except for reflect issues) and no other
code has been modified to make sure the changes are backwards
compatible.

Once this is accepted I'll pull the Agent version out.

### CODE REVIEW

File: api/renewer.go [hunk 1]
```diff
@@ -112,144 +149,113 @@ func (c *Client) NewRenewer(i *RenewerInput) (*Renewer, error) {
...
 		stopCh:  make(chan struct{}),
+
+		errLifetimeWatcherNotRenewable: ErrLifetimeWatcherNotRenewable,
+		errLifetimeWatcherNoSecretData: ErrLifetimeWatcherNoSecretData,
 	}, nil
 }
 
+// Deprecated: exists only for backwards compatibility. Identical to
+// NewLifetimeWatcher with the exception of some error cases.
+func (c *Client) NewRenewer(i *LifetimeWatcherInput) (*LifetimeWatcher, error) {
```
Comment (chrishoffman) [inline api/renewer.go]: Does this method need to set the RenewBehavior to the old default?

Comment (chrishoffman) [reply api/renewer.go]: Nevermind, I see it is done in `Renew()`.  It may make sense to just do it here and let Renew pass through to Start.

Comment (jefferai) [reply api/renewer.go]: Or in both?

Comment (jefferai) [reply api/renewer.go]: Fixed.

**Comment by briankassouf** [comment]
Want to do the agent change in this PR too? Just to validate the change?

**Comment by jefferai** [comment]
Sure, I can.

**Pull request merged**